clippy_lints/src/restricted_locales.rs

```diff
@@ -0,0 +1,408 @@
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for usage of unicode scripts other than explicitly allowed
+    /// by the lint config.
+    ///
+    /// This lint doesn't take into account non-text scripts such as `Unknown` and `Linear_A`.
+    /// It also ignores the `Common` script type.
+    /// While configuring, be sure to use official script names from [the list of supported scripts][supported_scripts]
+    /// and replace spaces with undescores.
+    ///
+    /// For example, `Syloti Nagri` must be represented as `"Syloti_Nagri`".
+    ///
+    /// [supported_scripts]: http://www.unicode.org/standard/supported.html
+    ///
+    /// **Why is this bad?** It may be not desired to have many different locales for
+    /// identifiers in the codebase.
+    ///
+    /// Note that if you only want to allow plain English, you might want to use
+    /// built-in `non_ascii_idents` lint instead.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust
+    /// // Assuming that `clippy.toml` contains the following line:
+    /// // allowed-locales = ["Latin", "Cyrillic"]
+    /// let counter = 10; // OK, latin is allowed.
+    /// let счётчик = 10; // OK, cyrillic is allowed.
+    /// let zähler = 10; // OK, it's still latin.
+    /// let カウンタ = 10; // Will spawn the lint.
+    /// ```
+    pub RESTRICTED_LOCALES,
+    pedantic,
+    "usage of non-allowed locales"
+}
+
+#[derive(Clone, Debug)]
+pub struct RestrictedLocales {
+    whitelist: FxHashSet<AllowedScript>,
+}
+
+impl RestrictedLocales {
+    pub fn new(whitelist: FxHashSet<AllowedScript>) -> Self {
+        Self { whitelist }
+    }
+}
+
+impl_lint_pass!(RestrictedLocales => [RESTRICTED_LOCALES]);
+
+impl<'tcx> LateLintPass<'tcx> for RestrictedLocales {
+    fn check_pat(&mut self, cx: &LateContext<'tcx>, pat: &'tcx Pat<'_>) {
+        let ident = if let PatKind::Binding(.., ident, _) = pat.kind {
+            ident
+        } else {
+            return;
+        };
+
+        for symbol in ident.name.as_str().chars() {
+            let script = symbol.script();
+            if let Some(as_allowed) = AllowedScript::from_script(script) {
+                if !self.whitelist.contains(&as_allowed) {
+                    span_lint(
+                        cx,
+                        RESTRICTED_LOCALES,
+                        ident.span,
+                        &format!(
+                            "identifier `{}` has locale that is not allowed: {}",
+                            ident.name,
+                            script.full_name()
+                        ),
+                    );
+                    // We don't want to spawn warning multiple times over a single identifier.
+                    break;
+                }
+            }
+        }
+    }
+}
+
+/// List of Unicode scripts that can be used to restrict usage.
+/// It does not contain several locales from the `Script` enum,
+/// such as `Common` or `Unknown`.
```

Comment:
Kinda feel like I'd rather use the same enum and filter out these variants when the lint object is constructed